### PR TITLE
[bugfix] catch particles on domain edges for Gadget_FOF data

### PR DESCRIPTION
**Pull request by brittonsmith**
## PR Summary

In this frontend, and likely others, particles on domain boundaries were being lost because the code to bump them in from the edges only existed in `_initialize_index` and none of the other functions in the io handler that read particles. This fixes that by implementing a class method on the data file for reading particle positions and doing the correction there rather than have three copies of the same code.

In addition, I implemented this with a cache since particle positions need to be read in for every field read. For the rather large dataset I was testing, this reduced the load time for a field from ~3.5s to ~0.5s (after the positions had been read in once). Of course, this will increase the steady-state memory use, but I think it's probably worth it (although I'm open to other opinions).

Like many bugs of this nature, it took a pretty large dataset (73 GB) to find it, so including it for a test is probably difficult.

I suspect that this affects all similar frontends (all the halo catalogs at least). I'm open to porting this fix around if that's desirable. I could also implement some sort of `correct_positions` function at a higher class level.

## PR Checklist

- [ ] Code passes flake8 checker
- [ ] New features are documented, with docstrings and narrative docs
- [ ] Adds a test for any bugs fixed. Adds tests for new features.

### CODE REVIEW

**Comment by ngoldbaum** [comment]
For the really big dataset you found this on what’s the memory cost of the cache?

**Comment by brittonsmith** [comment]
This dataset has ~14 million halos, which comes out to ~300 MB for the positions. That's not even a very big dataset by particle standards, so I could see not wanting to cache for things much large than this. That could either be implemented on a case by case basis or configurable somehow, or I don't know, I could ditch the caching.

For what it's worth, the bulk of the 73 GB of this dataset is storing member particles and fields for each halo.

**Comment by ngoldbaum** [comment]
Looks like the test failure is real.

I thought about this a bit last night and I think it's probably OK to keep the cache around if the speedup is really so significant. @MatthewTurk might want to weigh in as well. It might be nice to make that configurable somehow so there's a knob to turn it off (e.g. in a keyword argument passed to `load` or in a yt config option). We'd probably want to explicitly test both cases though.

I don't think you should be responsible for fixing this elsewhere. However, if you could identify where it's a problem elsewhere and open an issue describing how you'd fix it that would help a ton.

**Comment by MatthewTurk** [comment]
I'm not super into this method of caching, but it seems to be a workable solution.

One thing that I'd be a bit curious about is just how the particles on the domain boundaries get lost -- where do they stop getting selected?  I'm wondering if we should modify the cython routines to ensure that the right edge is considered inclusively for particles or not.

File: yt/frontends/halo_catalog/data_structures.py [hunk 1]
```diff
@@ -44,14 +44,68 @@ def _setup_filenames(self):
...
+            return None
+
+        # These are 32 bit numbers, so we give a little lee-way.
+        # Otherwise, for big sets of particles, we often will bump into the
+        # domain edges.  This helps alleviate that.
+        dle = self.ds.domain_left_edge.to('code_length').v
+        dre = self.ds.domain_right_edge.to('code_length').v
+        dx = 2. * np.finfo(np.float32).eps
+        pos = self._read_particle_positions(ptype, f=f)
+        np.clip(pos, dle + dx, dre - dx, pos)
```
Comment (MatthewTurk) [inline yt/frontends/halo_catalog/data_structures.py]: I admit I'm not totally following this, but are we respecting the "only read in these bounds" rules that we allow in the constructor to the dataset object?

File: yt/frontends/halo_catalog/data_structures.py [hunk 2]
```diff
@@ -44,14 +44,68 @@ def _setup_filenames(self):
...
+
+        if cache_pos:
+            self._coords[ptype] = pos
```
Comment (MatthewTurk) [inline yt/frontends/halo_catalog/data_structures.py]: One thing to keep an eye on here is that we're caching it, but it's entirely possible that in-place operations will happen on the arrays later on.  It might be a good idea when accessing things from the cache to ensure that they're always in the right units.

**Comment by qobilidop** [comment]
Just one suggestion, if a real dataset is too large, could there be an artificial dataset made to test these edge cases? Otherwise, it looks good to me.

File: yt/frontends/halo_catalog/data_structures.py (hunk 1 above)
Comment (brittonsmith) [reply yt/frontends/halo_catalog/data_structures.py]: I don't understand this comment. What are these rules you're referring to?

This clipping strategy was code that appeared in `_initialize_index` for each of the frontends here. It seemed like a good idea to apply the same logic to the actual positions, but maybe it doesn't matter.

File: yt/frontends/halo_catalog/data_structures.py (hunk 2 above)
Comment (brittonsmith) [reply yt/frontends/halo_catalog/data_structures.py]: I didn't think of that. There may `YTArrays` made with these changing all the units around and it would be difficult to know. Maybe this just won't work.

File: yt/frontends/halo_catalog/data_structures.py (hunk 1 above)
Comment (ngoldbaum) [reply yt/frontends/halo_catalog/data_structures.py]: I think he's referring to supplying a `bounding_box` keyword argument to `yt.load()`:

```
In [1]: ds.domain_left_edge
Out[1]: YTArray([ -547.51629822,  -557.30091064, -3675.19602173]) code_length

In [2]: ds.domain_right_edge
Out[2]: YTArray([ 579.153078  ,  472.81018799, 1591.69712036]) code_length

In [3]: ad = ds.all_data()

In [4]: ad['gas', 'density'].shape
Out[4]: (169240,)

In [6]: ds = yt.load('TipsyGalaxy/galaxy.00300', bounding_box=[ds.domain_left_edge/2, ds.domain_right_edge/2])

In [7]: ds.domain_left_edge
Out[7]: YTArray([ -273.75814911,  -278.65045532, -1837.59801086]) code_length

In [8]: ds.domain_right_edge
Out[8]: YTArray([289.576539  , 236.40509399, 795.84856018]) code_length

In [9]: ad = ds.all_data()

In [10]: ad['gas', 'density'].shape
Out[10]: (167728,)
```

**Comment by brittonsmith** [comment]
If memory serves, they were getting lost in the `selector.select_points(x, y, z, 0.0)` call because they were exactly on the right boundary. I think it's more proper continue with a periodic boundary that is inclusive on the left and exclusive on the right since it avoids the possibility of having two numerical values for a single physical location. I can see that becoming a problem somewhere. In addition, nudging particles inward as a general practice could result in a significant displacement from a particle's actual position. Rather, if the boundary is periodic then we should be shifting the particles to their true position within the domain. I am going to change this PR to do that. No one seems to like the cache, so I'm going to get rid of that, too.

**Comment by brittonsmith** [comment]
@qobilidop, I've added a couple tests using some fake halo catalogs, including a test where I put particles on the domain boundaries. I've confirmed that this fails prior to this PR.

File: yt/frontends/halo_catalog/tests/test_outputs.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+          'current_time': YTQuantity(1, 'yr'),
+          'domain_left_edge': YTArray(np.zeros(3), 'cm'),
+          'domain_right_edge': YTArray(np.ones(3), 'cm')}
+    save_as_dataset(
+        ds, filename, data,
+        field_types=ftypes, extra_attrs=extra_attrs)
+    return filename
+
+class HaloCatalogTest(TempDirTest):
+    def test_halo_catalog(self):
```
Comment (ngoldbaum) [inline yt/frontends/halo_catalog/tests/test_outputs.py]: These tests need to be decorated with `@requires_module('h5py')`

**Comment by ngoldbaum** [comment]
@yt-fido test this please

File: yt/frontends/halo_catalog/tests/test_outputs.py (hunk 3 above)
Comment (brittonsmith) [reply yt/frontends/halo_catalog/tests/test_outputs.py]: Right, I always forget this. Thanks!
